### PR TITLE
ref: Remove match_regex

### PR DESCRIPTION
**Pull request by untitaker**
Remove match-regex and replace it with something much simpler. This needs more tests. It is tested in Python, where I manually ran the tests and they worked fine.

### CODE REVIEW

File: relay-general/derive/src/lib.rs [hunk 1]
```diff
@@ -1033,3 +1041,54 @@ fn parse_variant_attributes(attrs: &[syn::Attribute]) -> VariantAttrs {
 fn is_newtype(variant: &synstructure::VariantInfo) -> bool {
     variant.bindings().len() == 1 && variant.bindings()[0].ast().ident.is_none()
 }
+
+fn parse_character_set(ident: &Ident, value: &str) -> TokenStream {
```
Comment (untitaker) [inline relay-general/derive/src/lib.rs:1045]: This is the kind of function that is fun to write (always feeling smart when writing trivial parsers), but not great to look back at.

File: relay-general/derive/src/lib.rs [hunk 2]
```diff
@@ -885,12 +885,20 @@ fn parse_field_attributes(
                                         panic!("Got non string literal for trim_whitespace");
                                     }
                                 }
-                            } else if ident == "match_regex" {
+                            } else if ident == "allow_characters" || ident == "deny_characters" {
```
Comment (jan-auer) [inline relay-general/derive/src/lib.rs]: Since we have `max_chars` already, let's shorten this to `allow_chars` and `deny_chars`.

File: relay-general/src/processor/attrs.rs [hunk 3]
```diff
@@ -212,6 +213,61 @@ pub struct FieldAttrs {
     pub retain: bool,
 }
 
+/// A set of characters allowed or denied for a (string) field.
+///
+/// Note that this field is generated in the derive, it can't be constructed easily in tests.
+#[derive(Clone, Copy)]
+pub struct CharacterSet {
+    /// Generated in derive for performance. Can be left out when set is created manually.
+    pub is_match: Option<fn(char) -> bool>,
```
Comment (jan-auer) [inline relay-general/src/processor/attrs.rs]: _nit_: What if you always assign a default function that returns `true` or `false`? I would not even allow creating this struct manually, ever.

File: relay-general/src/processor/attrs.rs [hunk 4]
```diff
@@ -212,6 +213,61 @@ pub struct FieldAttrs {
...
+    /// A set of ranges that are allowed/denied within the character set
+    pub ranges: &'static [RangeInclusive<char>],
+    /// Whether the character set is inverted
+    pub is_negative: bool,
```
Comment (jan-auer) [inline relay-general/src/processor/attrs.rs:226]: Do you need the `is_negative` flag just for matching or also for the JSON schema. If you only need it for matching, you can invert the condition in the function instead.

File: relay-general/src/store/schema.rs [hunk 5]
```diff
@@ -120,15 +90,17 @@ where
...
     meta: &mut Meta,
     state: &ProcessingState<'_>,
 ) -> ProcessingResult {
-    if let Some(ref regex_string) = state.attrs().match_regex {
-        if !get_regex(regex_string).is_match(value) {
-            meta.add_error(Error::invalid("invalid characters in string"));
-            return Err(ProcessingAction::DeleteValueSoft);
+    if let Some(ref character_set) = state.attrs().characters {
+        for c in value.chars() {
+            if !character_set.is_match(c) {
```
Comment (jan-auer) [inline relay-general/src/store/schema.rs]: You can "simplify" this to:

```rust
if value.chars().any(|c| !character_set.is_match(c)) {
  // add error here
}
```

This would probably be a good utility function on CharacterSet, taking an `&str`.

File: relay-general/src/processor/attrs.rs [hunk 6]
```diff
@@ -212,6 +213,61 @@ pub struct FieldAttrs {
...
+    pub fn deny(ranges: &'static [RangeInclusive<char>]) -> Self {
+        CharacterSet {
+            ranges,
+            is_match: None,
+            is_negative: true,
+        }
+    }
+
+    #[inline]
+    pub fn is_match(&self, c: char) -> bool {
```
Comment (jan-auer) [inline relay-general/src/processor/attrs.rs]: This is a slightly confusing function to me at the moment. If I'm reading the code correctly, then `is_match` means `is_invalid`. Would it make more sense to just generate a function `is_valid`?

File: relay-general/src/processor/attrs.rs (hunk 3 above)
Comment (untitaker) [reply relay-general/src/processor/attrs.rs]: answering this and the other questions: I need ranges and is_negative purely for jsonschema. `is_match` field alone is sufficient for basic functionality.

File: relay-general/src/processor/attrs.rs (hunk 6 above)
Comment (untitaker) [reply relay-general/src/processor/attrs.rs]: `is_match` means `is_valid`. There is a double-negation with `is_negative`, which is false if the character set is inclusive. I can remove this function though if you're suggesting to not make this constructible outside of derive (which tbh I am not sure of, it does mean you can't impl traits manually

File: relay-general/src/store/schema.rs (hunk 5 above)
Comment (untitaker) [reply relay-general/src/store/schema.rs]: wanted to show invalid character in error message so I left it like that (could've went with `Result<(), char>` as return type but that's too complicated.

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,7 @@
 **Internal**:
 
 - Emit more useful normalization meta data for invalid tags. ([#808](https://github.com/getsentry/relay/pull/808))
+- Internal refactoring such that validating of characters in tags no longer uses regexes internally. ([#814](https://github.com/getsentry/relay/pull/814))
```
Comment (jan-auer) [inline CHANGELOG.md]: This needs to go into the "Unreleased" section.

File: relay-general/src/store/schema.rs (hunk 5 above)
Comment (jan-auer) [reply relay-general/src/store/schema.rs]: I see, that makes sense. It would be nice if we could print the entire valid character set (similar to how often password fields tell you which characters are valid, not which character you entered wrong).

**Review by jan-auer — APPROVED** [review]

**Review by jan-auer — APPROVED** [review]
